Allow interop resolvers to return self handle

am11 opened the pull request:
When user-defined resolvers return self handle via `GetMainProgramHandle()`, we lookup the cached handle in mono and fail to find one because `internal_module` is not cached in `native_library_module_map`.

Fix is to test the resolver returned handle against self; before the map lookup (which happens under the lock).

Close #77985.

am11 commented:
@lambdageek, @vargaz, candidate of backport? (`GetMainProgramHandle` is a new API in 7.0)

wasm failures are preexisting: https://github.com/dotnet/runtime/issues/64188, https://github.com/dotnet/runtime/issues/76454, https://github.com/dotnet/runtime/issues/78017

lambdageek commented:
/backport to release/7.0

am11 commented:
@lambdageek, could you please run the Android pipeline (`/azp run runtime-extra-platforms`) against this PR? I think Android test runner will fail here as well. I'll then push a fix to verify it works.

lambdageek commented:
/azp run runtime-extra-platforms

akoeplinger commented:
On main we can also do `/azp run runtime-android` to only run the android jobs.
I've looked at the test results from the `runtime-extra-platforms` jobs that are already in and they are indeed the same as @am11 suspected.

In general we should try to avoid merging a backport for PRs that weren't yet merged to main, we'd probably have caught this earlier.

am11 commented:
@lambdageek, could you re-run the leg? The tests are passing locally now. For the internal call, we needed to continue looking up in the map.

lambdageek commented:
/azp run runtime-android

lambdageek commented:
@am11 I've been meaning to ask:  would it be possible to add a unit test for #77985? probably something using the remote executor would work, and looking for some C standard lib function (like `_malloc` ?) on non-Windows?

am11 commented:
For some reason, it wasn't failing the debug build `./build.sh mono+libs -os Android` with missing data type (`gboolean`).🤔 Fixed in 4156ab6.

@lambdageek, sure. I will add a case in resolver tests.

am11 commented:
```
  [ 17%] Building C object mono/mini/CMakeFiles/monosgen-objects.dir/__/metadata/sgen-mono.c.o
  /Users/adeel/projects/runtime/src/mono/mono/metadata/native-library.c:677:138: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
  netcore_resolve_with_dll_import_resolver_nofail (MonoAssemblyLoadContext *alc, MonoAssembly *assembly, const char *scope, guint32 flags, is_internal)
```

`Wimplicit-int` should be an error IMO.

am11 commented:
Added a test and rebased. I had to change the way we resolve self handle on Android: use `dlopen(NULL, RTLD_NOW)` instead of setting it to `RTLD_DEFAULT`, which [is null on 64-bit platforms](https://cs.android.com/android/platform/superproject/+/master:bionic/libc/include/dlfcn.h;l=70;bpv=1;bpt=1) and [`RTLD_NOW`](https://cs.android.com/android/platform/superproject/+/master:bionic/tests/dlfcn_test.cpp;l=76?q=dlfcn_test&ss=android%2Fplatform%2Fsuperproject) is what they use it for handle to `self`.

@lambdageek, please azp run again. 🙂

am11 committed:
Allow interop resolvers to return self handle

lambdageek commented:
/azp run runtime-android

am11 committed:
Disable new test on windows and monointerpreter

am11 commented:
Thank you. So Android legs were green. I have disabled the new test on Windows and monointerpreter:

* `free` is not availabe in the self image on Windows (which is OK and aligns with @lambdageek's expectation for the test)
* `monointerpreter` is not OK, but that is a preexisting problem. Calling free on self handle via P/Invoke breaks IL interpreter on all mono supported OS. We should fix it (separately; not part of this PR).

lambdageek commented on src/tests/Interop/NativeLibrary/MainProgramHandle/MainProgramHandleTests.cs:
```diff
@@ -0,0 +1,46 @@
+    {
+        try
+        {
+            free(s_handle);
```
Why do we free `s_handle` here? That seems like we're asking for trouble.  `free (IntPtr.Zero)` would also work (`free (NULL)` is a no-op call - so we would just test that the we got the right symbol, which is what we want).

I'm assuming the failure on the mono interpreter legs is somehow related (we need to keep the native handle around longer for some reason?)

lambdageek replied:
Makes sense, thanks!

am11 replied:
I'm trying with `getppid()`. Seems like it will work everywhere (test on osx-arm64 and android-arm64). Lets try that after this CI run completes.

am11 replied:
Nvm, it didn't help. IL interp needs to be fixed. :)

am11 committed:
Add new test to monointerpreter ExcludeList

lambdageek commented:
/azp run runtime-android

am11 committed:
Try test with getppid on all platforms

am11 committed:
Revert "Try test with getppid on all platforms"

This reverts commit de8eced8b79cafde878b22caf89887fbd44bb05d.

lambdageek commented:
/azp run runtime-android

lambdageek commented:
/backport to release/7.0

lambdageek commented:
/azp run runtime-androidemulator

lambdageek commented:
New test passed on Android emulator.  LGTM.  Thanks @am11
